File: spc_up/services/ingest/excel.py
"""Excel spreadsheet parsing for bank transactions."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal, InvalidOperation
from pathlib import Path
from typing import Any

from openpyxl import load_workbook

from spc_up.models.entities import MovimentacaoDirecao

_REQUIRED_COLUMNS = frozenset({"data", "valor", "descricao"})
_OPTIONAL_COLUMNS = frozenset({"tipo"})
_VALID_HEADERS = _REQUIRED_COLUMNS | _OPTIONAL_COLUMNS

_TIPO_ENTRADA = frozenset({"C", "CREDITO", "CRÉDITO", "ENTRADA"})
_TIPO_SAIDA = frozenset({"D", "DEBITO", "DÉBITO", "SAIDA", "SAÍDA"})


def _normalize_header(value: Any) -> str | None:
    if value is None:
        return None
    key = str(value).strip().lower()
    if key in _VALID_HEADERS:
        return key
    return None


def _parse_date(value: Any) -> date:
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    text = str(value).strip()
    for fmt in ("%d/%m/%Y", "%Y-%m-%d", "%d-%m-%Y"):
        try:
            return datetime.strptime(text, fmt).date()
        except ValueError:
            continue
    raise ValueError(f"Data inválida: {value!r}")


def _parse_decimal(value: Any) -> Decimal:
    if isinstance(value, Decimal):
        return value
    if isinstance(value, bool):
        raise ValueError(f"Valor inválido: {value!r}")
    if isinstance(value, (int, float)):
        return Decimal(str(value))
    text = str(value).strip()
    if not text:
        raise ValueError("Valor inválido: vazio")
    if "," in text and "." in text:
        text = text.replace(".", "").replace(",", ".")
    elif "," in text:
        text = text.replace(",", ".")
    try:
        return Decimal(text)
    except InvalidOperation as exc:
        raise ValueError(f"Valor inválido: {value!r}") from exc


def _direction_from_tipo(tipo: Any, amount: Decimal) -> tuple[str, Decimal]:
    if tipo is not None and str(tipo).strip():
        token = str(tipo).strip().upper()
        if token in _TIPO_ENTRADA:
            return MovimentacaoDirecao.ENTRADA.value, abs(amount)
        if token in _TIPO_SAIDA:
            return MovimentacaoDirecao.SAIDA.value, abs(amount)
        raise ValueError(f"Tipo inválido: {tipo!r}")
    if amount >= 0:
        return MovimentacaoDirecao.ENTRADA.value, abs(amount)
    return MovimentacaoDirecao.SAIDA.value, abs(amount)


def _row_is_empty(values: dict[str, Any]) -> bool:
    return all(value is None or str(value).strip() == "" for value in values.values())


def parse_excel(path: str | Path) -> list[dict[str, Any]]:
    """Parse an Excel file into normalized transaction rows (same shape as OFX ingest)."""
    workbook = load_workbook(Path(path), read_only=True, data_only=True)
    try:
        sheet = workbook.active
        if sheet is None:
            return []

        header_row: list[Any] | None = None
        column_index: dict[str, int] = {}
        rows: list[dict[str, Any]] = []

        for row in sheet.iter_rows(values_only=True):
            if header_row is None:
                header_row = list(row)
                for index, cell in enumerate(header_row):
                    name = _normalize_header(cell)
                    if name is not None:
                        column_index[name] = index
                missing = _REQUIRED_COLUMNS - column_index.keys()
                if missing:
                    labels = ", ".join(sorted(missing))
                    raise ValueError(f"Colunas obrigatórias ausentes: {labels}")
                continue

            values = {
                name: row[column_index[name]] if column_index[name] < len(row) else None
                for name in column_index
            }
            if _row_is_empty(values):
                continue

            amount = _parse_decimal(values["valor"])
            direcao, valor = _direction_from_tipo(values.get("tipo"), amount)
            descricao = str(values["descricao"] or "").strip()

            rows.append(
                {
                    "data_movimento": _parse_date(values["data"]),
                    "valor": valor,
                    "descricao_raw": descricao,
                    "direcao": direcao,
                    "nr_extrato_bancario": None,
                }
            )
        return rows
    finally:
        workbook.close()

File: packages/core/src/ingest/excel.ts
import { readFile } from "node:fs/promises";

import ExcelJS from "exceljs";

import { MOVIMENTACAO_DIRECAO, type ParsedTransactionRow } from "./types";

const REQUIRED_COLUMNS = new Set(["data", "valor", "descricao"]);
const VALID_HEADERS = new Set([...REQUIRED_COLUMNS, "tipo"]);

const TIPO_ENTRADA = new Set(["C", "CREDITO", "CRÉDITO", "ENTRADA"]);
const TIPO_SAIDA = new Set(["D", "DEBITO", "DÉBITO", "SAIDA", "SAÍDA"]);

function normalizeHeader(value: unknown): string | null {
  if (value == null) {
    return null;
  }
  const key = String(value).trim().toLowerCase();
  return VALID_HEADERS.has(key) ? key : null;
}

function parseDate(value: unknown): Date {
  if (value instanceof Date) {
    return value;
  }
  const text = String(value).trim();
  for (const [pattern, fn] of [
    [/^(\d{2})\/(\d{2})\/(\d{4})$/, (m: RegExpMatchArray) => new Date(+m[3]!, +m[2]! - 1, +m[1]!)],
    [/^(\d{4})-(\d{2})-(\d{2})$/, (m: RegExpMatchArray) => new Date(+m[1]!, +m[2]! - 1, +m[3]!)],
    [/^(\d{2})-(\d{2})-(\d{4})$/, (m: RegExpMatchArray) => new Date(+m[3]!, +m[2]! - 1, +m[1]!)],
  ] as const) {
    const match = text.match(pattern);
    if (match) {
      return fn(match);
    }
  }
  throw new Error(`Data inválida: ${String(value)}`);
}

function parseDecimal(value: unknown): number {
  if (typeof value === "number") {
    return value;
  }
  let text = String(value).trim();
  if (!text) {
    throw new Error("Valor inválido: vazio");
  }
  if (text.includes(",") && text.includes(".")) {
    text = text.replace(/\./g, "").replace(",", ".");
  } else if (text.includes(",")) {
    text = text.replace(",", ".");
  }
  const num = Number.parseFloat(text);
  if (Number.isNaN(num)) {
    throw new Error(`Valor inválido: ${String(value)}`);
  }
  return num;
}

function directionFromTipo(
  tipo: unknown,
  amount: number,
): { direcao: ParsedTransactionRow["direcao"]; valor: string } {
  if (tipo != null && String(tipo).trim()) {
    const token = String(tipo).trim().toUpperCase();
    if (TIPO_ENTRADA.has(token)) {
      return { direcao: MOVIMENTACAO_DIRECAO.ENTRADA, valor: Math.abs(amount).toFixed(2) };
    }
    if (TIPO_SAIDA.has(token)) {
      return { direcao: MOVIMENTACAO_DIRECAO.SAIDA, valor: Math.abs(amount).toFixed(2) };
    }
    throw new Error(`Tipo inválido: ${String(tipo)}`);
  }
  if (amount >= 0) {
    return { direcao: MOVIMENTACAO_DIRECAO.ENTRADA, valor: Math.abs(amount).toFixed(2) };
  }
  return { direcao: MOVIMENTACAO_DIRECAO.SAIDA, valor: Math.abs(amount).toFixed(2) };
}

function rowIsEmpty(values: Record<string, unknown>): boolean {
  return Object.values(values).every(
    (value) => value == null || String(value).trim() === "",
  );
}

async function parseWorkbook(workbook: ExcelJS.Workbook): Promise<ParsedTransactionRow[]> {
  const sheet = workbook.worksheets[0];
  if (sheet == null) {
    return [];
  }

  const columnIndex: Record<string, number> = {};
  const rows: ParsedTransactionRow[] = [];
  let headerReady = false;

  sheet.eachRow((row) => {
    const values = row.values as unknown[];
    const cells = values.slice(1);

    if (!headerReady) {
      for (let index = 0; index < cells.length; index += 1) {
        const name = normalizeHeader(cells[index]);
        if (name != null) {
          columnIndex[name] = index;
        }
      }
      const missing = [...REQUIRED_COLUMNS].filter((col) => !(col in columnIndex));
      if (missing.length > 0) {
        throw new Error(`Colunas obrigatórias ausentes: ${missing.sort().join(", ")}`);
      }
      headerReady = true;
      return;
    }

    const record: Record<string, unknown> = {};
    for (const [name, index] of Object.entries(columnIndex)) {
      record[name] = cells[index];
    }
    if (rowIsEmpty(record)) {
      return;
    }

    const amount = parseDecimal(record.valor);
    const { direcao, valor } = directionFromTipo(record.tipo, amount);

    rows.push({
      dataMovimento: parseDate(record.data),
      valor,
      descricaoRaw: String(record.descricao ?? "").trim(),
      direcao,
      nrExtratoBancario: null,
    });
  });

  return rows;
}

export async function parseExcelBuffer(buffer: Buffer): Promise<ParsedTransactionRow[]> {
  const workbook = new ExcelJS.Workbook();
  await workbook.xlsx.load(buffer as unknown as Parameters<ExcelJS.Workbook["xlsx"]["load"]>[0]);
  return parseWorkbook(workbook);
}

export async function parseExcel(pathOrBuffer: string | Buffer): Promise<ParsedTransactionRow[]> {
  const buffer =
    typeof pathOrBuffer === "string" ? await readFile(pathOrBuffer) : pathOrBuffer;
  return parseExcelBuffer(buffer);
}
